fake_quant: make qparams shape consistent

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38587 fake_quant: make qparams shape consistent**

Summary:

Before this diff, scale+zp were initialized to tensors
with a single dimension and 1 element, and then switched
to scalar tensors after the first forward.

This diff makes the shape stay consistent.  This should fix
an issue reported when saving/loading models, which crashes
on this inconsistent shape.

Test Plan:

```
python test/test_quantization.py TestFakeQuantizePerTensor.test_fake_quant_preserves_qparam_shapes_for_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21605532](https://our.internmc.facebook.com/intern/diff/D21605532)

vkuzo committed:
fake_quant: make qparams shape consistent

Summary:

Before this diff, scale+zp were initialized to tensors
with a single dimension and 1 element, and then switched
to scalar tensors after the first forward.

This diff makes the shape stay consistent.  This should fix
an issue reported when saving/loading models, which crashes
on this inconsistent shape.

Test Plan:

```
python test/test_quantization.py TestFakeQuantizePerTensor.test_fake_quant_preserves_qparam_shapes_for_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "fake_quant: make qparams shape consistent"


Summary:

Before this diff, scale+zp were initialized to tensors
with a single dimension and 1 element, and then switched
to scalar tensors after the first forward.

This diff makes the shape stay consistent.  This should fix
an issue reported when saving/loading models, which crashes
on this inconsistent shape.

Test Plan:

```
python test/test_quantization.py TestFakeQuantizePerTensor.test_fake_quant_preserves_qparam_shapes_for_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21605532](https://our.internmc.facebook.com/intern/diff/D21605532)

[ghstack-poisoned]

vkuzo committed:
Update on "fake_quant: make qparams shape consistent"


Summary:

Before this diff, scale+zp were initialized to tensors
with a single dimension and 1 element, and then switched
to scalar tensors after the first forward.

This diff makes the shape stay consistent.  This should fix
an issue reported when saving/loading models, which crashes
on this inconsistent shape.

Test Plan:

```
python test/test_quantization.py TestFakeQuantizePerTensor.test_fake_quant_preserves_qparam_shapes_for_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21605532](https://our.internmc.facebook.com/intern/diff/D21605532)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_workflow_module.py:
```diff
@@ -568,6 +568,34 @@ def test_fake_quant_control(self):
+        x = torch.rand(4, 4, 4, 4)
+        m(x)
+        scale_shape_after = m.linear.activation_post_process.scale.shape
+        zero_point_shape_after = m.linear.activation_post_process.zero_point.shape
```
Looks good, should we also check for per-channel quant (i.e the weights?)

vkuzo replied:
just double checked, we expect the per-channel params to change (for C > 1), so it did not have the issue solved by this diff

vkuzo committed:
Update on "fake_quant: make qparams shape consistent"


Summary:

Before this diff, scale+zp were initialized to tensors
with a single dimension and 1 element, and then switched
to scalar tensors after the first forward.

This diff makes the shape stay consistent.  This should fix
an issue reported when saving/loading models, which crashes
on this inconsistent shape.

Test Plan:

```
python test/test_quantization.py TestFakeQuantizePerTensor.test_fake_quant_preserves_qparam_shapes_for_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21605532](https://our.internmc.facebook.com/intern/diff/D21605532)

[ghstack-poisoned]
